extension/ts/background-scripts/background.js/telemetry/Telemetry.ts

```diff
@@ -1,82 +1,116 @@
-import Glean from "@mozilla/glean/webext";
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Glean from "@mozilla/glean/dist/webext";
+import errors from "./generated/errors";
+import infobar from "./generated/infobar";
+import metadata from "./generated/metadata";
+import performance from "./generated/performance";
+import service from "./generated/service";
+import test from "./generated/test";
 // @ts-ignore
 import { custom } from "./generated/pings";
 // @ts-ignore
-import { to_lang, from_lang } from "./generated/metadata";
-
-Glean.initialize("bergamot-extension", true, { debug: { logPings: true } });
+import { toLang, fromLang } from "./generated/metadata";
 
-const _set_meta = (lang_from: string, lang_to: string) => {
-  to_lang.set(lang_to);
-  from_lang.set(lang_from);
+const generatedGleanModules = {
+  errors,
+  infobar,
+  metadata,
+  performance,
+  service,
+  test,
 };
 
-const _dynamic_call = (
-  category: string,
-  name: string,
-  method: string,
-  value: any = null,
-) => {
-  import(`./generated/${category}`)
-    .then(module => {
-      if (value != null) {
-        module[name][method](value);
+type MetricsCategory =
+  | "errors"
+  | "infobar"
+  | "metadata"
+  | "performance"
+  | "service"
+  | "test";
+
+type GleanCustomPingSubmitResults = undefined;
+
+export class Telemetry {
```

Comment:
why class is better than functions? this stuff should be callable globally from random places without creating objects. We might use singleton, but I think it's just overcomplication comparing to functions

Reply:
Pure functions without side effects are fine (where each function is stored in it's own file), but I'd prefer to keep anything that relies on something that should be initialized in classes. It also makes it easier to swap out whole chunks of functionality in dependent tests (eg any code that sends telemetry can use TestTelemetry instead of Telemetry). Classes also allows us to define which parts of the interface is private and public in a more idiomatic way, but that is more an aesthetic reason. All in all, to keep consistency with the rest of the code in the extension, I'd prefer if we use classes. What do you think?

Reply:
I'm ok with OOP in general, but for vertically integrated things like logging, metrics etc I prefer to use global objects and functions if possible. You're right, dependencies on objects are way easier to test, make stubs etc. The problem is usability: we'll need to instantiate this class everywhere and pass the instance everywhere or use a singleton which is essentially the same as global object. We can use static methods maybe. For testing we just should be able to provide a method which enables/disables uploading

Reply:
Agree. And since Glean is already uses singletons everywhere and don't need much wrapping, we might just want to use the generated methods directly and not even have a Telemetry.js file. I trust you'll find a good setup wither way :)